File: my-folder/0017-letter-combinations-of-a-phone-number/solution.cpp
class Solution {
public:
    void recursion(string digits,unordered_map<char,string> key,vector<string> &ans,string temp,int i){
        if(i==digits.size()){
            ans.push_back(temp);
            return;
        }
        string s = key[digits[i]];
        for(char c:s){
            recursion(digits,key,ans,temp+c,i+1);
        }
    }
    vector<string> letterCombinations(string digits) {
        if(digits.size()==0) return {};
        unordered_map<char,string> keyboard = {
            {'2', "abc"},
            {'3', "def"},
            {'4', "ghi"},
            {'5', "jkl"},
            {'6', "mno"},
            {'7', "pqrs"},
            {'8', "tuv"},
            {'9', "wxyz"}
        };
        vector<string> ans;
        recursion(digits,keyboard,ans,"",0);
        return ans;
       

        
    }
};

File: my-folder/0017-letter-combinations-of-a-phone-number/solution.java
class Solution {
    void getCombinations(String digits,List<String> ans, Map<Character,String> mp,String temp,int ind){
        if(digits.length()==temp.length()){
            ans.add(temp);
            return;
        }
        String s = mp.get(digits.charAt(ind));
        for(char c:s.toCharArray()){
            getCombinations(digits,ans,mp,temp+c,ind+1);
        }
    }
    public List<String> letterCombinations(String digits) {
        if (digits == null || digits.length() == 0) {
            return new ArrayList<>();
        }
        Map<Character,String> mp = new HashMap<>();
        mp.put('2',"abc");
        mp.put('3',"def");
        mp.put('4',"ghi");
        mp.put('5',"jkl");
        mp.put('6',"mno");
        mp.put('7',"pqrs");
        mp.put('8',"tuv");
        mp.put('9',"wxyz"); 
        List<String> ans = new ArrayList<>();
        getCombinations(digits,ans,mp,new String(),0);
        return ans;
        
    }
}
